variantlib/validators.py

```diff
@@ -153,3 +158,96 @@ def validate_type(value: Any, expected_type: type) -> None:
     wrong_type = _validate_type(value, expected_type)
     if wrong_type is not None:
         raise ValidationError(f"Expected {expected_type}, got {wrong_type}")
+
+
+def validate_variant_json(data: dict) -> None:
+    """
+    Validate the variant JSON data structure.
+    """
+
+    if not isinstance(data, dict):
+        raise ValidationError(f"Expected a dictionary for data, got {type(data)}")
+
+    if VARIANTS_JSON_VARIANT_DATA_KEY not in data:
+        raise ValidationError(f"Key `variants` is missing in data: {list(data.keys())}")
+
+    variant_data = data[VARIANTS_JSON_VARIANT_DATA_KEY]
+
+    if not isinstance(variant_data, dict):
+        raise ValidationError(
+            f"Expected a dictionary for `data['variants']`, got `{type(variant_data)}`."
+        )
+
+    hash_re = re.compile(rf"^[0-9a-f]{{{VARIANT_HASH_LEN}}}$")
+    namespace_re = re.compile(VALIDATION_NAMESPACE_REGEX)
+    feature_re = re.compile(VALIDATION_FEATURE_REGEX)
+    value_re = re.compile(VALIDATION_VALUE_REGEX)
```

Comment:
Perhaps we should keep compiled versions of these in `variantlib.constants`.

Reply:
True - Doing that - We should do that to other regex too